### PR TITLE
FIX: make Fresnel and Fraunhofer backward the true inverse of forward

### PR DESCRIPTION
**Pull request by stevehenke**
`propagate_backward` multiplied by `C1C2 / C0` and by `B`, applying both unit-modulus phase screens twice in the forward direction instead of undoing them. Round-trip error was 1.385 (Fresnel) and 0.453 (Fraunhofer); it is now 3e-16. Both classes also wrapped a bare `fft2` around kernels built on a centered grid, so the far field came back in corner order; they are now centered in, centered out, matching the angular spectrum and the `fft_shift=False` convention used at finite propagation distances. Forward output moves by half an array in both axes as a result.

Angular spectrum: clamp `1 - ratio` before the square root, since the NaN it produced above the cutoff survived `torch.where` as a gradient and poisoned d(loss)/d(distance) when slice spacings are optimizable; block evanescent modes instead of passing them through unattenuated; conjugate the transfer function in `propagate_backward` rather than dividing by it, which is equivalent where |tf| == 1 and stays defined where it is 0.

Propagator selection now compares conjugate-plane pitch against source pitch per axis using |z|. It previously tested a signed Fresnel number against a positive threshold, so every negative distance chose the Fresnel transform, and the single `sqrt(N*M)` threshold could alias an anisotropic geometry on its narrow axis.

Also drop the per-call `F2` buffer re-registration, give Fraunhofer the double-precision kernels and buffer registration its sibling already had, and correct a stale comment about DataParallel.

Add tests/test_propagate.py, covering round trip, per-class operator identity, energy, selection, evanescent handling, distance gradients, and analytic references (Gaussian beam waist, amplitude and Gouy phase; rect aperture against the Dirichlet kernel). 32 of its 70 tests fail against the previous implementation.

# Features/fixes

Describe new features or bugs fixed.

# Related issues (optional)

Add links to issues related to this pull request, if any.

# Mentions

Mention team members responsible for reviewing this change. 

# Checklist (Don't skip!)

To check a checklist item, change the space in the backets with an `x`, *i.e.*, `[ ] -> [x]`.

Have you...
- [ ] Formatted your code properly adhering to PEP-8? Considering using RUFF to format your code automatically (https://docs.astral.sh/ruff/installation/).
- [ ] Cleaned up your code to remove large commented-out blocks?
- [ ] Properly documented any new API knobs and functions with non-trivial signatures?
- [ ] Resolved all merge conflicts with the main branch?
- [ ] Checked the diffs between your branch and the main branch to ensure that your changes are not introducing any regressions or unnecessary/irrelevant changes?
- [ ] Checked the diffs to ensure your PR only includes one feature/fix? Multiple features/fixes/changes should be split into different PRs.
- [ ] Added unit tests for new features if necessary?

### CODE REVIEW

File: tests/test_propagate.py [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+def test_distance_gradient_is_nonzero_without_evanescent_modes():
+    """Control: the clamp must not flatten the ordinary gradient."""
+
+    def parameters_factory(propagation_distance_m):
+        return _params(propagation_distance_m, wavelength_m=1e-10, pixel_width_m=1e-6)
+
+    gradient = _distance_gradient(parameters_factory)
+
+    assert torch.isfinite(gradient)
+    assert float(gradient) != 0.0
```
Comment (mdw771) [inline tests/test_propagate.py:480]: Since the "loss" here is the total power, it is independent of the distance, so the exact gradient should be 0; as such, we shouldn't assert that it is nonzero here. The test passes probably because there are still some non-zero residual due to numerical precision. We should probably remove this test; the finite gradient in the presence of evanescent modes is tested elsewhere.

**Review by mdw771 — COMMENTED** [review]
Thanks a lot Steve! This is very helpful. I didn't realized that we've been multiplying the shifted phase factor A with unshifted FFT of wavefield * B in the Fresnel transform propagator in the past; I'm surprised that I didn't encounter a major problem with near field reconstruction. 
Just one comment about a test; we might want to remove it. If that sounds reasonable to you I'll remove that test and merge. 
I also delivered another fix that prevents the single-FT propagators from breaking on negative propagation distances.

File: tests/test_propagate.py (hunk 1 above)
Comment (stevehenke) [reply tests/test_propagate.py]: Sounds good to me. Thanks, Ming!

**Review by mdw771 — APPROVED** [review]
